### PR TITLE
Assets - Add bock asset features to compile dependencies and version - WIP

### PR DESCRIPTION
**Pull request by jesseeproductions**
Adds methods to set dependencies and version based on complied block assets. 

Utilize the following methods when adding an asset:
Set path automatically:
```
Asset::add(
	'event-schedule-manager-speaker-list-block-js',
	'js/app/speaker-list/index.js'
	)
	->set_as_compiled()
	->add_to_group( 'event-schedule-manager-editor' )
	->register();
```
Or set the path manually:
```
Asset::add(
	'event-schedule-manager-speaker-list-block-js',
	'speaker-list-block-index.js'
	)
	->set_as_compiled()
        ->set_compiled_path( 'js/app/speaker-list/index' )
	->add_to_group( 'event-schedule-manager-editor' )
	->register();
```

Both examples assume the default asset path is in `src/resources/`

### CODE REVIEW

File: src/Assets/Asset.php [hunk 1]
```diff
@@ -217,6 +217,16 @@ class Asset {
...
 	protected ?string $version = null;
 
+	/**
+	 * The compiled data for all assets already loaded.
+	 *
+	 * @since TBD
+	 *
+	 * @var array
+	 */
+	protected array $compiled_data = [];
```
Comment (bordoni) [inline src/Assets/Asset.php:227]: I would call it just `compiled` and avoid the `_data` suffix.

File: src/Assets/Asset.php [hunk 2]
```diff
@@ -1258,4 +1268,101 @@ public function set_type( string $type ) {
...
+	 * Gets the data for a given compiled Asset.
+	 *
+	 * @since TBD
+	 *
+	 * @param string $src The partial path to the asset.
+	 *
+	 * @return ?array
+	 */
+	public function get_compiled_data( string $src ): ?array {
+		$file = $this->get_root_path() . $this->get_path() . $src . '.asset.php';
```
Comment (bordoni) [inline src/Assets/Asset.php:1282]: Avoid passing a `$src` here, use the `$this->slug`. Which would change how the rest of this method works a bit. But it should be only one compiled variable per `Asset` instance.

File: src/Assets/Asset.php [hunk 3]
```diff
@@ -1258,4 +1268,101 @@ public function set_type( string $type ) {
...
+		}
+
+		if ( ! empty( $this->dependencies ) ) {
+			$block_dependencies = array_merge( $block_dependencies, $this->dependencies );
+		}
+
+		$this->set_dependencies( ...$block_dependencies );
+
+		return $this;
+	}
```
Comment (bordoni) [inline src/Assets/Asset.php:1367]: I would avoid all of these methods, and add a conditional method like `is_compiled()` that determines if the Dependencies and version should come from a dependency compiled include.

**Review by bordoni — CHANGES_REQUESTED** [review]
I would love to see us simplify to allow assets to be compiled or not and then just add the compiled variables to the ones that we have set.

File: src/Assets/Asset.php (hunk 2 above)
Comment (jesseeproductions) [reply src/Assets/Asset.php]: I removed this method.

File: src/Assets/Asset.php (hunk 3 above)
Comment (jesseeproductions) [reply src/Assets/Asset.php]: I removed these methods.

**Comment by jesseeproductions** [comment]
- I updated the coding to have set_compiled_data() method that can be used to setup the asset to use the compiled data if found. 
- I did pass the src to it because the slug would have to be changed to the src and gives you less options in naming.
- The get_version and get_dependencies are updated to use compiled data if found.

File: src/Assets/Asset.php [hunk 4]
```diff
@@ -481,6 +491,13 @@ public function get_condition() {
...
 	 */
 	public function get_dependencies(): array {
+		if (
+			$this->is_compiled()
+			&& ! empty( $this->compiled['dependencies'] )
+		) {
+			return $this->compiled['dependencies'];
+		}
+
 		return $this->dependencies;
```
Comment (bordoni) [inline src/Assets/Asset.php:501]: Lets make sure we merge and `array_unique` these.

**Review by bordoni — CHANGES_REQUESTED** [review]

**Comment by borkweb** [comment]
I'm closing this one in favor of a further evolution in #19

**Pull request closed without merging**